File: src/LCR084.7p8L0Z.1000235/solution.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 */
var permuteUnique = function (nums) {
    nums.sort((a, b) => a - b);
    const used = new Array(nums.length).fill(false);
    const sequence = [];
    const result = [];
    const backTracking = () => {
        if (sequence.length === nums.length) {
            result.push(sequence.slice());
            return;
        }
        for (let i = 0; i < nums.length; i++) {
            if (used[i] || (i > 0 && nums[i] === nums[i - 1] && !used[i - 1])) {
                continue;
            }
            used[i] = true;
            sequence.push(nums[i]);
            backTracking();
            sequence.pop();
            used[i] = false;
        }
    };
    backTracking();
    return result;
};

File: src/LCR084.7p8L0Z.1000235/solution.java
class Solution {
    public List<List<Integer>> permuteUnique(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> result = new ArrayList<>();
        backTracking(nums,new boolean[nums.length],new ArrayList<>(),result );
        return result;
    }
    
    private void backTracking(int[] nums,boolean[] used,List<Integer> accumulate,List<List<Integer>> result){
        if(accumulate.size() == nums.length){
            result.add(new ArrayList<>(accumulate));
            return;
        }
        
        for(int i=0;i<nums.length;i++){
            if(used[i] || (i>0 && nums[i] == nums[i-1] && !used[i-1])){
                continue;
            }
            used[i] = true;
            accumulate.addLast(nums[i]);
            backTracking(nums,used,accumulate,result);
            accumulate.removeLast();
            used[i] = false;
        }
        
        
    }
    
}